secureboot_policy.go

```diff
@@ -696,88 +696,89 @@ type secureBootDbSet struct {
+	c.firmwareVerificationEvents = make(tpm2.DigestList, len(p.firmwareVerificationEvents))
+	copy(c.firmwareVerificationEvents, p.firmwareVerificationEvents)
+	c.shimVerificationEvents = make(tpm2.DigestList, len(p.shimVerificationEvents))
+	copy(c.shimVerificationEvents, p.shimVerificationEvents)
 
-	n.dbUpdateLevel = p.dbUpdateLevel
-	n.dbSet = p.dbSet
+	p.subBranches = append(p.subBranches, c)
 
-	n.firmwareVerificationEvents = make(tpm2.DigestList, len(p.firmwareVerificationEvents))
-	copy(n.firmwareVerificationEvents, p.firmwareVerificationEvents)
-	n.shimVerificationEvents = make(tpm2.DigestList, len(p.shimVerificationEvents))
-	copy(n.shimVerificationEvents, p.shimVerificationEvents)
+	return c
+}
 
-	return n
+type sbLoadEventAndBranches struct {
+	event    *EFIImageLoadEvent
+	branches []*secureBootPolicyGenBranch
 }
 
-type loadEventAndPaths struct {
-	event *EFIImageLoadEvent
-	paths []*secureBootPolicyGenPath
+func (e *sbLoadEventAndBranches) branch(event *EFIImageLoadEvent) *sbLoadEventAndBranches {
+	var branches []*secureBootPolicyGenBranch
+	for _, b := range e.branches {
+		if b.profile == nil {
+			continue
+		}
+		branches = append(branches, b.branch())
+	}
+	return &sbLoadEventAndBranches{event, branches}
 }
 
 // secureBootPolicyGen is the main structure involved with computing secure boot policy PCR digests.
 type secureBootPolicyGen struct {
 	*EFISecureBootPolicyProfileParams
 }
 
-// extendMeasurement extends the supplied digest to the current value of pcrValue for the specified event path.
-func (g *secureBootPolicyGen) extendMeasurement(path *secureBootPolicyGenPath, digest tpm2.Digest) {
-	path.extendMeasurement(g.PCRAlgorithm, digest)
+// extendMeasurement extends the supplied digest in to the specified branch.
+func (g *secureBootPolicyGen) extendMeasurement(branch *secureBootPolicyGenBranch, digest tpm2.Digest) {
+	if len(branch.subBranches) > 0 {
+		panic("This branch has already been branched")
+	}
+	branch.profile.ExtendPCR(g.PCRAlgorithm, secureBootPCR, digest)
 }
 
-// extendFirmwareVerificationMeasurement extends the supplied digest to the current value of pcrValue for the specified event path,
-// and records the extended digest in order to avoid measuring the same verification event more than once.
-func (g *secureBootPolicyGen) extendFirmwareVerificationMeasurement(path *secureBootPolicyGenPath, digest tpm2.Digest) {
-	path.extendVerificationMeasurement(g.PCRAlgorithm, digest, Firmware)
+// extendVerificationMeasurement extends the supplied digest and records that the digest has been measured by the specified source in
+// ths specified branch.
+func (g *secureBootPolicyGen) extendVerificationMeasurement(branch *secureBootPolicyGenBranch, digest tpm2.Digest, source EFIImageLoadEventSource) {
+	var digests *tpm2.DigestList
+	switch source {
+	case Firmware:
+		digests = &branch.firmwareVerificationEvents
+	case Shim:
+		digests = &branch.shimVerificationEvents
+	}
+	*digests = append(*digests, digest)
```

Comment:
this kind of manipulating of things on branch make it for sure look like it would best live on *Branch itself

Reply:
I've moved this entire method now.